src/union.ts

```diff
@@ -337,19 +337,88 @@ export class Union {
    * @param fs the filesystem interface to be added to the queue of FS's
    * @returns this instance of a unionFS
    */
-  use(fs: IFS): this {
-    this.fss.push(fs);
+  use(fs: IFS, options: VolOptions = {}): this {
+    this.fss.push([this.createFS(fs, options), options]);
     return this;
   }
 
+  /**
+   * At the time of the [[use]] call, we create our sync, async and promise methods
+   * for performance reasons
+   *
+   * @param fs
+   * @param options
+   */
+  private createFS(fs: IFS, { readable = true, writeable = true }: VolOptions): IFS {
+    const noop = (..._args: any[]) => {
+      throw new SkipMethodError();
+    };
+    const createFunc = (method: string): any => {
+      if (!fs[method])
+        return (...args: any[]) => {
+          throw new Error(`Method not supported: "${method}" with args "${args}"`);
+        };
+      return (...args: any[]) => fs[method as string].apply(fs, args);
+    };
+
+    return {
+      ...fs,
+      ...fsSyncMethodsRead.reduce((acc, method) => {
+        acc[method] = readable ? createFunc(method) : noop;
+        return acc;
+      }, {} as Record<typeof fsSyncMethodsRead[number], ReturnType<typeof createFunc>>),
```

Comment:
These casts are not needed, as TypeScript will infer the type.

For future, `reduce` takes a generic parameter that should be used since casting actually weakens the type :)

i.e

```
[1, 2, 3].reduce<number[]>((a, s) => a.concat(s * 2), [])
```

Reply:
i spent a few minutes on this trying the best way using generics etc. I think if we're weren't using the initialValues param generics would be better yeah

[TS playground](https://www.typescriptlang.org/play/?ssl=6&ssc=57&pln=6&pc=44#code/MYewdgzgLgBADgQzFCMC8MDaByRyLYA0MuSKATNgLowKqiRQCwAUFAJ5wCmMACmQCUuoAE4ATVBiGixAHg7cQAM3hkImMAFcAtgCMuIqsSTsAfK1YNoMANZgAlsBsGIARgBcfQcJDjJq-AA6ES4xTWAuAApIhGBgY35kBLIASTAxLgAPAEp0UxgAb1o4zAREqBoMAAM8FAASArLU9KyAXyqAbhgQqE0RMGLgGFbiAtbsy3BrO0dnEQhyT3LpXwl0AJRg0PCuWWWfP1No2PjacuTkNIycvMLB0vLKmBq1BqbLlsz2rp6+gZPhqNxpNGLYHE4XABmJbeGT+WoQLZhCLHOIXKDoq5ZXJofJFE4PMhPF74N7lLFfTrdLi9fqDQGFVq0VD7OETFisIA)

Reply:
Yeah that's because with the type your using [the object must have the properties defined](https://www.typescriptlang.org/play/?ssl=1&ssc=1&pln=8&pc=1#code/MYewdgzgLgBADgQzFCMC8MDaByRyLYA0MuSKATNgLowKqiRQCwAUFAJ5wCmMACmQCUuoAE4ATVBiGixAHg7cQAM3hkImMAFcAtgCMuIqsSTsAfK1YNoMANZgAlsBsGIARgBcfQcJDjJq-AA6ES4xTWAuAApIhGBgY35kBLIASTAxLgAPAEp0UxgAb1o4zAREqBoMAAM8FAASArLU9KyAXyqAbhgQqE0RMGLgGFbiAtbsy3BrO0dnEQhyT3LpXwl0AJRg0PCuWWWfP1No2PjacuTkNIycvMLB0vLKmBq1BqbLlsz2rp6+gZPhqMSLUCJ5NC0lPYwKFiKR8JQwRCoaFhhMWFZYDMnC4AMxLbwyfwgrZhCLHOIXKCUq5ZXJofJFE4PMhPF74N7lGlfTrdLi9fqDQGFVq0VD7Qlo1hAA) :)

Hence why it's a good idea to just drop the type all together given that it's inferable.

Reply:
The cast isn't dangerous here - TS wouldn't allow it otherwise. Surely you see no. 3 is simplest [here](https://www.typescriptlang.org/play/?ssl=8&ssc=15&pln=8&pc=27#code/MYewdgzgLgBADgQzFCMC8MDaByRyLYA0MuSKATNgLowKqiRQCwAUFAJ5wCmMACmQCUuoAE4ATVBiGixAHg7cQAM3hkImMAFcAtgCMuIqsSTsAfK1YNoMANZgAlsBsGIARgBcfQcJDjJq-AA6ES4xTWAuAApIhGBgY35kBLIASTAxLgAPAEp0UxgAb1o4zAREqBoMAAM8FAASArLU9KyAXyqAbhgQqE0RMGLgGFbiAtbsy3BrO0dnEQhyT3LpXwl0AJRg0PCuWWWfP1No2PjacuTkNIycvMLB0vLKmBq1BqbLlsz2rp6+gZPhqMSLUCJ5NC0lPYwKFiKR8JQwRCoaFhhMWFZYDMnC4AMxLbwyfwgrZhCLHOIXKCUq5ZXJofJFE4PMhPF74N7lGlfTrdLi9fqDQGFVq0VD7QlojG2BzY+YAFnWxJCpKiMQpZzI1M+dIZ93eFXWbPqjU5n2+vP5-ziQrGkqmmJlcwgAFZFWoSTs9gTVhAjmrTvqtdcdXcmfrWSCOc1rubfgKASNhWigA)

It's not inferable (and i wouldn't be bothering otherwise) - i need the result to not be string indexed

Reply:
> The cast isn't dangerous here - TS wouldn't allow it otherwise. 

Actually when casting like this Typescript will warn about excess properties but *not* missing required fields, which is exactly what you're seeing here.

That's why we've made a few [linting rules](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/consistent-type-assertions.md#objectliteraltypeassertions) to catch [this pattern](https://github.com/typescript-eslint/typescript-eslint/pull/1707).

> Surely you see no. 3 is simplest here

Actually I see them all as being pretty complicated.

> i need the result to indexed by the list elements

I'm not quite understanding why; what errors is TypeScript giving you?

But then if that is the case, the solution is to provide all of those properties; that's what TypeScript is trying to tell you: you're missing properties :)

Reply:
because we're spreading the initial fs is doesn't complain without them so have removed